File: web/js/tetromino.js
export const BOARD_ROWS = 20;
export const BOARD_COLS = 10;

export const TETROMINO_TYPES = ['I', 'O', 'T', 'S', 'Z', 'J', 'L'];

export const TETROMINO_COLORS = {
  I: '#00d9f2',
  O: '#f2d900',
  T: '#bf00d9',
  S: '#00d933',
  Z: '#f22626',
  J: '#0059f2',
  L: '#f28c00',
};

export const BASE_OFFSETS = {
  I: [[0, 0], [0, 1], [0, 2], [0, 3]],
  O: [[0, 0], [0, 1], [1, 0], [1, 1]],
  T: [[0, 1], [1, 0], [1, 1], [1, 2]],
  S: [[0, 1], [0, 2], [1, 0], [1, 1]],
  Z: [[0, 0], [0, 1], [1, 1], [1, 2]],
  J: [[0, 0], [1, 0], [1, 1], [1, 2]],
  L: [[0, 2], [1, 0], [1, 1], [1, 2]],
};

export function randomType() {
  return TETROMINO_TYPES[Math.floor(Math.random() * TETROMINO_TYPES.length)];
}

function rotatePoint(row, col, times) {
  let r = row;
  let c = col;
  for (let i = 0; i < times % 4; i += 1) {
    const newRow = c;
    const newCol = -r;
    r = newRow;
    c = newCol;
  }
  return { row: r, col: c };
}

export function getOffsets(type, rotation) {
  const base = BASE_OFFSETS[type];
  if (type === 'O') return base;

  const centerRow = Math.floor(base.reduce((sum, [r]) => sum + r, 0) / base.length);
  const centerCol = Math.floor(base.reduce((sum, [, c]) => sum + c, 0) / base.length);

  return base.map(([row, col]) => {
    const translatedRow = row - centerRow;
    const translatedCol = col - centerCol;
    const rotated = rotatePoint(translatedRow, translatedCol, rotation);
    return [rotated.row + centerRow, rotated.col + centerCol];
  });
}

export function createPiece(type, row = 0, col = BOARD_COLS / 2 - 2, rotation = 0) {
  return { type, row, col, rotation };
}

export function movePiece(piece, dRow, dCol) {
  return { ...piece, row: piece.row + dRow, col: piece.col + dCol };
}

export function rotatePiece(piece, clockwise = true) {
  return {
    ...piece,
    rotation: clockwise ? (piece.rotation + 1) % 4 : (piece.rotation + 3) % 4,
  };
}

export function absolutePositions(piece) {
  return getOffsets(piece.type, piece.rotation).map(([r, c]) => [piece.row + r, piece.col + c]);
}

File: NamchonTetris/NamchonTetris/Models/Tetromino.swift
import SwiftUI

enum TetrominoType: CaseIterable, Equatable {
    case I, O, T, S, Z, J, L

    var color: Color {
        switch self {
        case .I: return Color(red: 0.0, green: 0.85, blue: 0.95)
        case .O: return Color(red: 0.95, green: 0.85, blue: 0.0)
        case .T: return Color(red: 0.75, green: 0.0, blue: 0.85)
        case .S: return Color(red: 0.0, green: 0.85, blue: 0.2)
        case .Z: return Color(red: 0.95, green: 0.15, blue: 0.15)
        case .J: return Color(red: 0.0, green: 0.35, blue: 0.95)
        case .L: return Color(red: 0.95, green: 0.55, blue: 0.0)
        }
    }

    /// Block offsets relative to the piece origin (row, col).
    var baseOffsets: [(Int, Int)] {
        switch self {
        case .I: return [(0, 0), (0, 1), (0, 2), (0, 3)]
        case .O: return [(0, 0), (0, 1), (1, 0), (1, 1)]
        case .T: return [(0, 1), (1, 0), (1, 1), (1, 2)]
        case .S: return [(0, 1), (0, 2), (1, 0), (1, 1)]
        case .Z: return [(0, 0), (0, 1), (1, 1), (1, 2)]
        case .J: return [(0, 0), (1, 0), (1, 1), (1, 2)]
        case .L: return [(0, 2), (1, 0), (1, 1), (1, 2)]
        }
    }

    static func random() -> TetrominoType {
        allCases.randomElement()!
    }
}

struct ActivePiece: Equatable {
    var type: TetrominoType
    var row: Int
    var col: Int
    var rotation: Int

    var offsets: [(Int, Int)] {
        let base = type.baseOffsets
        guard type != .O else { return base }

        let centerRow = base.map(\.0).reduce(0, +) / base.count
        let centerCol = base.map(\.1).reduce(0, +) / base.count

        return base.map { (row, col) in
            let translatedRow = row - centerRow
            let translatedCol = col - centerCol
            let rotated = rotatePoint(row: translatedRow, col: translatedCol, times: rotation)
            return (rotated.row + centerRow, rotated.col + centerCol)
        }
    }

    func absolutePositions() -> [(Int, Int)] {
        offsets.map { (row + $0.0, col + $0.1) }
    }

    func moved(byRow dRow: Int, byCol dCol: Int) -> ActivePiece {
        var piece = self
        piece.row += dRow
        piece.col += dCol
        return piece
    }

    func rotated(clockwise: Bool = true) -> ActivePiece {
        var piece = self
        if clockwise {
            piece.rotation = (rotation + 1) % 4
        } else {
            piece.rotation = (rotation + 3) % 4
        }
        return piece
    }

    private func rotatePoint(row: Int, col: Int, times: Int) -> (row: Int, col: Int) {
        var r = row
        var c = col
        for _ in 0..<(times % 4) {
            let newRow = c
            let newCol = -r
            r = newRow
            c = newCol
        }
        return (r, c)
    }
}
